src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+
+
+/**
+ * Handle all data, assume that the first row is the first reading (skip this row).
+ * @Example
+ * 	row 0: reading #0
+ *  row 1: reading #1
+ * 	row 2: reading #2
+ * => reading #1 = row 1
+ *    reading #2 = row 2
+ *    reading #0 may be the cumulative value from unknown readings that may or may not have been inserted before
+ * @param {object[[]]} rows
+ * @param {number} meterID meter id being input
+ * @param {boolean} isCumulative true if the data is cumulative
+ * @param {boolean} cumulativeReset true if the cumulative data can reset
+ * @param {string} resetStart a string representation of the start time a cumulativeReset may occur after. The default resetStart time is 0:00:00
+ * @param {string} resetEnd a string representation of the end time a cumulativeReset may occur before. The default resetEnd time is 23:59:99
+ * @param readingRepetition value is 1 if reading is not duplicated. 2 if repeated twice and so on (E-mon D-mon meters)
+ * @param {boolean} onlyEndTime true if the data only has an endTimestamp
+ * @param {number} Tgap the allowed time variation in milliseconds that a gap may occur between two readings 
+ * @param {number} Tlen the allowed time variation in milliseconds that two readings may deviate from each other
+ * @param {dict} conditionSet used to validate readings (minVal, maxVal, minDate, maxDate, interval, maxError)
+ */
+
+function processData(rows, meterID, isCumulative, cumulativeReset, resetStart = '0:00:00.000', 
+					resetEnd = '23:59:99.999', readingRepetition, onlyEndTime = false, Tgap, Tlen, conditionSet) {
+
+	// If processData is succesfully finished then return result = [R0, R1, R2...RN]
+	const result = [];
+
+	let errMsg = '';
+
+	let meterReading = 0;
+	let meterReading1 = 0;
+	let meterReading2 = 0;
+
+	let startTimestamp = moment(0);
+	let endTimestamp = moment(0);
+
+	let currentReading = new Reading(meterID, -1, startTimestamp, endTimestamp);
+	let prevReading = currentReading;
+	let readingOK = true;
+	for (let index = readingRepetition; index <= rows.length; ++index) {
+		// To read data where same reading is repeated. Like E-mon D-mon meters
+		if ((index - readingRepetition) % readingRepetition === 0) {
+			if (onlyEndTime) {
+				// The startTimestamp of this reading is the endTimestamp of the previous reading
+				startTimestamp = prevReading.endTimestamp;
+				endTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			else {
+				startTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+				endTimestamp = moment(rows[index - readingRepetition][2], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			if (isCumulative && isFirst(prevReading.endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing with cumulative data.';
+			}
+			if (onlyEndTime && isFirst(prevReading.endTimestamp)) {
```

Comment:
Done and noted. The way it's done now means that if the data is the first ever meter readings passed in for cumulative data and onlyEndTimes are used then we append the first drop error message so the user sees: 

"The first reading must be dropped when dealing with cumulative data. The first reading must be dropped when dealing with only endTimestamps."

I've also considered potentially just using the first one so that user doesn't see the second sentence but I think it makes more sense that the user sees as much information as possible. This can be changed if deemed necessary.